libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -2278,21 +2278,27 @@ void ExpressionCompiler::appendAndOrOperatorCode(BinaryOperation const& _binaryO
 
 void ExpressionCompiler::appendCompareOperatorCode(Token _operator, Type const& _type)
 {
-	solAssert(_type.sizeOnStack() == 1, "Comparison of multi-slot types.");
 	if (_operator == Token::Equal || _operator == Token::NotEqual)
 	{
-		if (FunctionType const* funType = dynamic_cast<decltype(funType)>(&_type))
+		FunctionType const* funType = dynamic_cast<decltype(funType)>(&_type);
+		if (funType && funType->kind() == FunctionType::Kind::External)
 		{
-			if (funType->kind() == FunctionType::Kind::Internal)
+			m_context << Instruction::SWAP3;
+			m_context << Instruction::EQ;
+			m_context << Instruction::SWAP2;
+			m_context << Instruction::EQ;
+			m_context << Instruction::AND;
+		}
+		else
+		{
+			if (funType && funType->kind() == FunctionType::Kind::Internal)
 			{
-				// We have to remove the upper bits (construction time value) because they might
-				// be "unknown" in one of the operands and not in the other.
 				m_context << ((u256(1) << 32) - 1) << Instruction::AND;
 				m_context << Instruction::SWAP1;
 				m_context << ((u256(1) << 32) - 1) << Instruction::AND;
```

Comment:
For taking care of the dirty higher-order bits you should be able to do something similar to this here, both for the selector and the address parts. The selector is 4 bytes, i.e. you can mask with ``((u256(1) << 32) - 1)``, the address part is 160 bits, i.e. you can mask with ``((u256(1) << 160) - 1)``.
I also just noticed that you removed the comment above the masking of internal function types, was that intentional?